src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -2691,6 +2703,25 @@ internal bool SupportsCompilation([NotNullWhen(false)] out string? reason)
             reason = null;
             return true;
 
+            static bool HasCaseInsensitiveBackReferences(RegexNode node)
+            {
+                if (node.Kind is RegexNodeKind.Backreference && (node.Options & RegexOptions.IgnoreCase) != 0)
+                {
+                    return true;
+                }
+
+                int childCount = node.ChildCount();
+                for (int i = 0; i < childCount; i++)
+                {
+                    if (HasCaseInsensitiveBackReferences(node.Child(i)))
```

Comment:
This is using recursion without any guards for stack depth.

That might be ok if you move it to after the ExceedsMaxDepthAllowedDepth check.